examples/app/vsgshadertoy/vsgshadertoy.cpp

```diff
@@ -0,0 +1,332 @@
+    auto bindDescriptorSet = vsg::BindDescriptorSet::create(VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, descriptorSet);
+
+    // create StateGroup as the root of the scene/command graph to hold the GraphicsPipeline, and binding of Descriptors to decorate the whole graph
+    auto node = vsg::StateGroup::create();
+    node->add(bindGraphicsPipeline);
+    node->add(bindDescriptorSet);
+
+    // set up vertex and index arrays
+    auto vertices = vsg::vec2Array::create(
+        {{-1.0f, -1.0f},
+         {1.0f, -1.0f},
+         {1.0f, 1.0f},
+         {-1.0f, 1.0f}}); // VK_FORMAT_R32G32_SFLOAT, VK_VERTEX_INPUT_RATE_INSTANCE, VK_BUFFER_USAGE_VERTEX_BUFFER_BIT, VK_SHARING_MODE_EXCLUSIVE
+
+    auto texcoords = vsg::vec2Array::create(
+        {{0.0f, 1.0f},
+         {1.0f, 1.0f},
+         {1.0f, 0.0f},
+         {0.0f, 0.0f}}); // VK_FORMAT_R32G32_SFLOAT, VK_VERTEX_INPUT_RATE_VERTEX, VK_BUFFER_USAGE_VERTEX_BUFFER_BIT, VK_SHARING_MODE_EXCLUSIVE
+
+    auto indices = vsg::ushortArray::create(
+        {0, 1, 2,
+         2, 3, 0}); // VK_BUFFER_USAGE_INDEX_BUFFER_BIT, VK_SHARING_MODE_EXCLUSIVE
+
+    // setup geometry
+    auto drawCommands = vsg::Commands::create();
+    drawCommands->addChild(vsg::BindVertexBuffers::create(0, vsg::DataList{vertices, texcoords}));
+    drawCommands->addChild(vsg::BindIndexBuffer::create(indices));
+    drawCommands->addChild(vsg::DrawIndexed::create(6, 1, 0, 0, 0));
+
+    // add drawCommands to transform
+    node->addChild(drawCommands);
+
+    return node;
+}
+
+class MouseHandler : public vsg::Inherit<vsg::Visitor, MouseHandler>
+{
+public:
+    void apply(vsg::PointerEvent& pointerEvent) override
+    {
+        lastPointerEvent = &pointerEvent;
+    }
+    
+
+    vsg::ref_ptr<vsg::PointerEvent> lastPointerEvent;
+};
+
+int main(int argc, char** argv)
+{
+    // set up defaults and read command line arguments to override them
+    vsg::CommandLine arguments(&argc, argv);
+
+    auto options = vsg::Options::create();
+    options->fileCache = vsg::getEnv("VSG_FILE_CACHE");
+    options->paths = vsg::getEnvPaths("VSG_FILE_PATH");
+
+    auto windowTraits = vsg::WindowTraits::create();
+    windowTraits->debugLayer = arguments.read({"--debug", "-d"});
+    windowTraits->apiDumpLayer = arguments.read({"--api", "-a"});
+    arguments.read({"--window", "-w"}, windowTraits->width, windowTraits->height);
+
+    if (arguments.errors()) return arguments.writeErrorMessages(std::cerr);
+
+
+    windowTraits->height = 600;
+    windowTraits->width = 800;
+
+    auto scene = vsg::Group::create();
+
+    // Read the shader from the command line
+    std::string toyShader;
+    if (argc < 2)
+    {
+        windowTraits->windowTitle = "vsgshadertoy";
+        toyShader = defaultShader;
+    }
+    else {
+        toyShader = readFile(argv[1]);
+        string title;
+        // extract the filename from the path
+        auto pos = string(argv[1]).find_last_of("/");
```

Comment:
Would this work on Windows (which uses '\')? Might be better to use std::filesystem's filename() function? (std::filesystem is C++17 so should be fine with VSG?)

Reply:
You're right. Won't work. I'll fix it.

Reply:
std::filesystem is officially part of C++17 so "should" be fine, except it's gcc and clang have only recently got there act together and properly support it out of the box, older versions don't properly support std::filesystem, it's why I had to resort to writing a vsg::Path class that is equivalent to std::filesystem::path.

The VSG has various path manipulator convenience functions:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/io/FileSystem.h
https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/io/Path.h#L218

Ironically these are more easily portable than the mess that was/is std::filesystem.

Reply:
Ok, I rewrote the code to use vsg::Path() instead of std::last_of("/")